src/transformStateWithClones.js

```diff
@@ -6,8 +6,43 @@
  *
  * @return {Object[]}
  */
+
+const addProperties = (state, properties, res) => {
+  Object.assign(state, properties);
+  res.push({ ...state });
+};
+
+const removeProperty = (state, keysToRemove, res = []) => {
+  keysToRemove.forEach((key) => {
+    delete state[key];
+  });
+  res.push({ ...state });
+};
+
+const clearState = (state, res) => {
+  Object.keys(state).forEach((key) => delete state[key]);
+
+  res.push({ ...state });
+};
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = { ...state };
+  const res = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'addProperties':
+        return addProperties(newState, action.extraData, res);
+
+      case 'removeProperties':
+        return removeProperty(newState, action.keysToRemove, res);
+
+      default:
+        return clearState(newState, res);
```

Comment:
- You should clear only in case of action type is `clear`
- In the default case you can throw an error, so someone who use your function can understand that they pass incorrect type

Reply:
The comments have been corrected.